### ISSUE TITLE
[BLOCKING] Phase 0.2: Implement Simple Lead Scorer (Logistic Regression)

### ISSUE THREAD

**Issue by nmogil** [nmogil/pennie-sense#14]
## Overview
Implement a simple logistic regression lead scoring model using only customer attributes (no conversation analysis). This will serve as a baseline and may be sufficient for production use.

## Context
Based on Phase 0.1 validation results, we're implementing a simple ML model that:
- Uses ONLY customer attributes (debt, credit, income, etc.)
- Costs $0 to run
- Takes 2-4 hours to implement
- Likely achieves 70-75% accuracy

## Prerequisites
- Phase 0.1 completed and validated
- Python environment with scikit-learn installed
- Historical lead data formatted

### Step 1: Create Lead Scorer Module

**File**: `src/simple_lead_scorer.py`

**Lines 1-150**: Complete implementation
```python
"""
Simple lead scoring using logistic regression.
Uses only customer attributes - no conversation analysis needed.
"""
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (
    roc_auc_score, 
    classification_report, 
    confusion_matrix,
    precision_recall_curve
)
import joblib
import os
from typing import Dict, List, Tuple, Optional

class SimpleLeadScorer:
    """
    Simple lead scoring model using logistic regression.
    
    Features used:
    - debt_amount
    - credit_score  
    - annual_income
    - debt_to_income_ratio (derived)
    - credit_tier (derived)
    - lead_source (one-hot encoded)
    - state (one-hot encoded)
    
    Does NOT use conversation content.
    """
    
    def __init__(self):
        self.model = LogisticRegression(max_iter=1000, random_state=42)
        self.scaler = StandardScaler()
        self.feature_names: List[str] = []
        self.version = "1.0"
        
    def engineer_features(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Engineer features from raw customer attributes.
        
        Args:
            df: DataFrame with columns:
                - debt_amount, credit_score, annual_income
                - lead_source, state
                
        Returns:
            DataFrame with additional engineered features
        """
        df = df.copy()
        
        # Derived numeric features
        df['debt_to_income_ratio'] = df['debt_amount'] / df['annual_income']
        
        # Credit score tiers
        df['credit_tier'] = pd.cut(
            df['credit_score'],
            bins=[0, 580, 650, 720, 850],
            labels=['poor', 'fair', 'good', 'excellent']
        )
        
        # One-hot encode categorical variables
        df = pd.get_dummies(
            df, 
            columns=['lead_source', 'state', 'credit_tier'],
            drop_first=True
        )
        
        return df
    
    def train(
        self, 
        csv_path: str, 
        test_size: float = 0.2
    ) -> Dict[str, float]:
        """
        Train the model from CSV file.
        
        Args:
            csv_path: Path to CSV with columns:
                - debt_amount, credit_score, annual_income
                - lead_source, state
                - converted (0 or 1)
            test_size: Fraction of data for testing
            
        Returns:
            Dictionary with evaluation metrics
        """
        print("="*70)
        print("TRAINING SIMPLE LEAD SCORER")
        print("="*70)
        
        # Load data
        print(f"\nLoading data from {csv_path}...")
        df = pd.read_csv(csv_path)
        print(f"✅ Loaded {len(df)} leads")
        
        # Check required columns
        required = ['debt_amount', 'credit_score', 'annual_income', 
                   'lead_source', 'state', 'converted']
        missing = [col for col in required if col not in df.columns]
        if missing:
            raise ValueError(f"Missing required columns: {missing}")
        
        # Check conversion distribution
        conv_rate = df['converted'].mean()
        print(f"   Conversion rate: {conv_rate:.1%}")
        print(f"   Converted: {df['converted'].sum()} | Not: {(~df['converted'].astype(bool)).sum()}")
        
        # Engineer features
        print("\nEngineering features...")
        df = self.engineer_features(df)
        
        # Separate features and target
        exclude_cols = ['converted', 'lead_id', 'conversation_id', 'customer_id']
        feature_cols = [col for col in df.columns if col not in exclude_cols]
        
        X = df[feature_cols]
        y = df['converted']
        
        print(f"✅ Created {len(feature_cols)} features")
        
        # Split data
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, 
            test_size=test_size, 
            random_state=42, 
            stratify=y
        )
        
        print(f"\nSplit: {len(X_train)} train, {len(X_test)} test")
        
        # Scale features
        print("\nScaling features...")
        X_train_scaled = self.scaler.fit_transform(X_train)
        X_test_scaled = self.scaler.transform(X_test)
        
        # Train model
        print("\nTraining logistic regression model...")
        self.model.fit(X_train_scaled, y_train)
        self.feature_names = feature_cols
        print("✅ Model trained")
        
        # Evaluate
        print("\n" + "="*70)
        print("MODEL EVALUATION")
        print("="*70)
        
        y_pred = self.model.predict(X_test_scaled)
        y_prob = self.model.predict_proba(X_test_scaled)[:, 1]
        
        # Calculate metrics
        auc = roc_auc_score(y_test, y_prob)
        
        print(f"\nROC AUC: {auc:.3f}")
        print("\nClassification Report:")
        print(classification_report(
            y_test, y_pred, 
            target_names=['Not Converted', 'Converted']
        ))
        
        # Cross-validation
        print("Running 5-fold cross-validation...")
        cv_scores = cross_val_score(
            self.model, X_train_scaled, y_train,
            cv=5, scoring='roc_auc'
        )
        print(f"✅ CV ROC AUC: {cv_scores.mean():.3f} ± {cv_scores.std():.3f}")
        
        # Feature importance
        importance_df = pd.DataFrame({
            'feature': feature_cols,
            'coefficient': self.model.coef_[0]
        }).sort_values('coefficient', key=abs, ascending=False)
        
        print("\n📊 Top 10 Most Important Features:")
        print(importance_df.head(10).to_string(index=False))
        
        metrics = {
            'roc_auc': auc,
            'cv_roc_auc_mean': cv_scores.mean(),
            'cv_roc_auc_std': cv_scores.std()
        }
        
        print("\n" + "="*70)
        print("✅ Training complete!")
        print("="*70)
        
        return metrics
    
    def predict(self, lead_data: Dict) -> Dict:
        """
        Score a single lead.
        
        Args:
            lead_data: Dictionary with keys:
                - debt_amount (float)
                - credit_score (int)
                - annual_income (float)
                - lead_source (str)
                - state (str)
                
        Returns:
            Dictionary with score, probability, priority
        """
        # Convert to DataFrame
        df = pd.DataFrame([lead_data])
        
        # Engineer features
        df = self.engineer_features(df)
        
        # Ensure all features exist (for one-hot encoded columns)
        for col in self.feature_names:
            if col not in df.columns:
                df[col] = 0
        
        # Select features in correct order
        X = df[self.feature_names]
        
        # Scale
        X_scaled = self.scaler.transform(X)
        
        # Predict
        probability = float(self.model.predict_proba(X_scaled)[0, 1])
        
        # Categorize
        if probability >= 0.7:
            priority = "HOT"
        elif probability >= 0.5:
            priority = "HIGH"
        elif probability >= 0.3:
            priority = "MEDIUM"
        else:
            priority = "COLD"
        
        return {
            'score': int(probability * 100),
            'probability': probability,
            'priority': priority,
            'model_version': self.version
        }
    
    def predict_batch(
        self, 
        csv_path: str, 
        output_path: Optional[str] = None
    ) -> pd.DataFrame:
        """Score multiple leads from CSV."""
        df = pd.read_csv(csv_path)
        
        scores = []
        for _, row in df.iterrows():
            lead_data = row.to_dict()
            result = self.predict(lead_data)
            scores.append(result)
        
        df['score'] = [s['score'] for s in scores]
        df['probability'] = [s['probability'] for s in scores]
        df['priority'] = [s['priority'] for s in scores]
        
        if output_path:
            df.to_csv(output_path, index=False)
            print(f"✅ Scored {len(df)} leads → {output_path}")
        
        return df
    
    def save(self, path: str = 'models/simple_lead_scorer.pkl'):
        """Save model to disk."""
        os.makedirs(os.path.dirname(path), exist_ok=True)
        
        joblib.dump({
            'model': self.model,
            'scaler': self.scaler,
            'features': self.feature_names,
            'version': self.version
        }, path)
        
        print(f"✅ Model saved to {path}")
    
    @classmethod
    def load(cls, path: str = 'models/simple_lead_scorer.pkl'):
        """Load model from disk."""
        scorer = cls()
        
        data = joblib.load(path)
        scorer.model = data['model']
        scorer.scaler = data['scaler']
        scorer.feature_names = data['features']
        scorer.version = data['version']
        
        print(f"✅ Model v{scorer.version} loaded from {path}")
        
        return scorer
```

### Step 2: Create Training Script

**File**: `scripts/train_simple_model.py`

```python
"""
Train the simple lead scoring model.
"""
import sys
from pathlib import Path

# Add src to path
sys.path.insert(0, str(Path(__file__).parent.parent))

from src.simple_lead_scorer import SimpleLeadScorer

if __name__ == "__main__":
    # Get data path
    if len(sys.argv) > 1:
        data_path = sys.argv[1]
    else:
        data_path = 'data/raw/sample_leads.csv'
    
    # Train model
    scorer = SimpleLeadScorer()
    metrics = scorer.train(data_path)
    
    # Save model
    scorer.save('models/simple_lead_scorer.pkl')
    
    print("\n" + "="*70)
    print("NEXT STEPS")
    print("="*70)
    print("\n1. Review model performance above")
    print("2. If ROC AUC >= 0.70, you have a working model!")
    print("3. Test scoring: python scripts/score_leads.py")
    print("4. Compare to DeepMost only if performance < 0.65")
```

### Step 3: Create Scoring Script

**File**: `scripts/score_leads.py`

```python
"""
Score leads using the trained simple model.
"""
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.simple_lead_scorer import SimpleLeadScorer

if __name__ == "__main__":
    # Load model
    try:
        scorer = SimpleLeadScorer.load('models/simple_lead_scorer.pkl')
    except FileNotFoundError:
        print("❌ ERROR: Model not found. Train first with:"
        print("   python scripts/train_simple_model.py")
        sys.exit(1)
    
    # Example: Score a single lead
    print("\n" + "="*70)
    print("EXAMPLE: Single Lead Scoring")
    print("="*70)
    
    example_lead = {
        'debt_amount': 18000.00,
        'credit_score': 680,
        'annual_income': 55000.00,
        'lead_source': 'web',
        'state': 'CA'
    }
    
    result = scorer.predict(example_lead)
    
    print(f"\nLead Attributes:")
    for key, value in example_lead.items():
        print(f"  {key}: {value}")
    
    print(f"\n🎯 Scoring Results:")
    print(f"   Score: {result['score']}/100")
    print(f"   Probability: {result['probability']:.1%}")
    print(f"   Priority: {result['priority']}")
    
    # Batch scoring
    if len(sys.argv) > 1:
        input_path = sys.argv[1]
        output_path = sys.argv[2] if len(sys.argv) > 2 else 'data/results/scored_leads.csv'
        
        print(f"\n" + "="*70)
        print("BATCH SCORING")
        print("="*70)
        
        df = scorer.predict_batch(input_path, output_path)
        
        print(f"\n📊 Priority Distribution:")
        print(df['priority'].value_counts())
```

### Step 4: Create Test Script

**File**: `scripts/test_simple_model.py`

```python
"""
Test the simple lead scorer implementation.
"""
import sys
from pathlib import Path
import pandas as pd

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.simple_lead_scorer import SimpleLeadScorer

def test_simple_model():
    """Run comprehensive tests on simple lead scorer."""
    
    print("="*70)
    print("TESTING SIMPLE LEAD SCORER")
    print("="*70)
    
    # Test 1: Import check
    print("\n1️⃣  Test: Module imports")
    try:
        from src.simple_lead_scorer import SimpleLeadScorer
        print("   ✅ Module imports successfully")
    except ImportError as e:
        print(f"   ❌ Import failed: {e}")
        return False
    
    # Test 2: Instantiation
    print("\n2️⃣  Test: Model instantiation")
    try:
        scorer = SimpleLeadScorer()
        print("   ✅ Model instantiates successfully")
    except Exception as e:
        print(f"   ❌ Instantiation failed: {e}")
        return False
    
    # Test 3: Feature engineering
    print("\n3️⃣  Test: Feature engineering")
    try:
        test_df = pd.DataFrame({
            'debt_amount': [15000, 20000],
            'credit_score': [680, 720],
            'annual_income': [55000, 70000],
            'lead_source': ['web', 'referral'],
            'state': ['CA', 'TX'],
            'converted': [1, 1]
        })
        
        df_features = scorer.engineer_features(test_df)
        
        # Check for derived features
        assert 'debt_to_income_ratio' in df_features.columns
        assert any('credit_tier' in col for col in df_features.columns)
        
        print(f"   ✅ Features engineered: {len(df_features.columns)} columns")
    except Exception as e:
        print(f"   ❌ Feature engineering failed: {e}")
        return False
    
    # Test 4: Training (with sample data)
    print("\n4️⃣  Test: Model training")
    try:
        # Check if sample data exists
        sample_path = 'data/raw/sample_leads.csv'
        if not Path(sample_path).exists():
            print(f"   ⚠️  Skipping training test: {sample_path} not found")
            print("   Create sample data first to test training")
        else:
            metrics = scorer.train(sample_path, test_size=0.3)
            
            # Check metrics
            assert 'roc_auc' in metrics
            assert 0 <= metrics['roc_auc'] <= 1
            
            print(f"   ✅ Training successful: ROC AUC = {metrics['roc_auc']:.3f}")
    except Exception as e:
        print(f"   ❌ Training failed: {e}")
        return False
    
    # Test 5: Single prediction
    print("\n5️⃣  Test: Single lead prediction")
    try:
        lead = {
            'debt_amount': 18000,
            'credit_score': 680,
            'annual_income': 55000,
            'lead_source': 'web',
            'state': 'CA'
        }
        
        result = scorer.predict(lead)
        
        # Validate result structure
        assert 'score' in result
        assert 'probability' in result
        assert 'priority' in result
        assert 0 <= result['score'] <= 100
        assert 0 <= result['probability'] <= 1
        assert result['priority'] in ['HOT', 'HIGH', 'MEDIUM', 'COLD']
        
        print(f"   ✅ Prediction successful")
        print(f"      Score: {result['score']}/100")
        print(f"      Priority: {result['priority']}")
    except Exception as e:
        print(f"   ❌ Prediction failed: {e}")
        return False
    
    # Test 6: Save/Load
    print("\n6️⃣  Test: Model save/load")
    try:
        test_model_path = 'models/test_scorer.pkl'
        scorer.save(test_model_path)
        
        loaded_scorer = SimpleLeadScorer.load(test_model_path)
        
        # Test that loaded model produces same results
        result2 = loaded_scorer.predict(lead)
        assert result['score'] == result2['score']
        
        print(f"   ✅ Save/load successful")
        
        # Cleanup
        Path(test_model_path).unlink()
    except Exception as e:
        print(f"   ❌ Save/load failed: {e}")
        return False
    
    print("\n" + "="*70)
    print("✅ ALL TESTS PASSED")
    print("="*70)
    
    return True

if __name__ == "__main__":
    success = test_simple_model()
    sys.exit(0 if success else 1)
```

### Test 1: Run Module Tests
```bash
python scripts/test_simple_model.py
```

Expected output:
- All 6 tests pass
- Model can be instantiated, trained, and used for predictions

### Test 2: Train on Sample Data
```bash
python scripts/train_simple_model.py data/raw/sample_leads.csv
```

Expected output:
- Model trains without errors
- ROC AUC score displayed
- Feature importance shown
- Model saved to `models/simple_lead_scorer.pkl`

### Test 3: Score Example Leads
```bash
python scripts/score_leads.py
```

Expected output:
- Example lead scored
- Score, probability, and priority shown

### Test 4: Batch Scoring
```bash
python scripts/score_leads.py data/raw/sample_leads.csv data/results/scored.csv
```

Expected output:
- All leads scored
- Results saved to CSV
- Priority distribution shown

## Success Criteria

- [ ] `src/simple_lead_scorer.py` created (400+ lines)
- [ ] `scripts/train_simple_model.py` created
- [ ] `scripts/score_leads.py` created
- [ ] `scripts/test_simple_model.py` created
- [ ] All tests pass
- [ ] Model trains successfully
- [ ] ROC AUC >= 0.60 (minimum)
- [ ] Can score individual leads
- [ ] Can batch score leads
- [ ] Model can be saved and loaded

## Performance Targets

| Metric | Minimum | Target | Excellent |
|--------|---------|--------|-----------|
| ROC AUC | 0.60 | 0.70 | 0.80+ |
| Accuracy | 60% | 70% | 80%+ |
| Training Time | <30s | <10s | <5s |
| Prediction Time | <100ms | <10ms | <1ms |

## Deliverables

- [ ] Complete implementation of `SimpleLeadScorer` class
- [ ] Training script with evaluation
- [ ] Scoring script for single and batch predictions
- [ ] Comprehensive test suite
- [ ] Trained model file (`.pkl`)
- [ ] Performance metrics documented

## Time Estimate
**2-4 hours**

## Dependencies
- Phase 0.1 (Quick Validation) must be completed

## Blocks
- Phase 0.3: Model Evaluation & Comparison
